### PR TITLE
Local copy first

### PR DESCRIPTION
**Pull request by TurboGit**
Recently discussed on the mailing-list. This makes using the local-copy first and then make it possible to have fast workflow on SSD drive even though the original file is still accessible.

### CODE REVIEW

File: src/common/image.c [hunk 1]
```diff
@@ -185,13 +185,16 @@ void dt_image_full_path(const int imgid, char *pathname, size_t pathname_len, gb
   }
   sqlite3_finalize(stmt);
 
-  if(*from_cache && !g_file_test(pathname, G_FILE_TEST_EXISTS))
+  if(*from_cache)
   {
-    _image_local_copy_full_path(imgid, pathname, pathname_len);
-    *from_cache = TRUE;
+    char lc_pathname[pathname_len];
```
Comment (LebedevRI) [inline src/common/image.c]: This should rather be
```
char lc_pathname[PATH_MAX] = { 0 };
_image_local_copy_full_path(imgid, lc_pathname, sizeof(lc_pathname));
```

**Comment by LebedevRI** [comment]
I must say, personally, i'm not convinced about this yet.
What about performance penalty? Any measurable difference?

File: src/common/image.c (hunk 1 above)
Comment (TurboGit) [inline src/common/image.c]: Argg!!! Indeed, I won't get this right :)

**Comment by TurboGit** [comment]
No difference as in fact there is no more IO involved, just reversed check.

**Comment by TurboGit** [comment]
Also, there is loose just wins in this. No slow-down and potential huge speed up by using the cached images on SSD or standard drive instead of slow remote drives like NAS.

**Comment by mgehre** [comment]
This feature would be really nice to have, because its exactly my workflow (storage on NAS, local copy for editing on local SSD)

**Comment by nsrosenqvist** [comment]
+1, I use network based file storage as well

**Comment by LebedevRI** [comment]
Why isn't this merged yet? What is missing, a review?

**Comment by TurboGit** [comment]
Yes a review would be nice.

**Comment by pedrocr** [comment]
This code doesn't seem too easy to follow and has unfixed bugs:

http://redmine.darktable.org/issues/10602

I think there's some room for refactoring and even redesign of the feature instead of piling on more stuff.

**Comment by TurboGit** [comment]
@pedrocr : Sorry I disagre, dt does not support moving around images outside of it's interface. It can be blamed if people insist on playing bad trick with it. Do you have a proposal to avoid such situation?

**Comment by pedrocr** [comment]
That's not the bug, the bug is that the warning should never be about the local copy but the original. If the original exists there's no issue and if the original and the copy don't exist the problem should be the original not existing. Caches should be deletable with no issues. The bug is minor it just made me look at the code a little time ago.

**Comment by TurboGit** [comment]
@pedrocr, I see but this is actually fixed by this patch that's why I did not really understand the concern about this issue. Indeed as we now first look for the local copy, the last check is done for the original image and so the error message properly reference it.

**Comment by pedrocr** [comment]
@TurboGit I haven't looked at this code so if it fixes that issue as well all the better.

**Comment by TurboGit** [comment]
Ok, let's review it and merge it ok.

**Pull request merged**